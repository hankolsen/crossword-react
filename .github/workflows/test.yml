name: melodikryzzet-tests

on: 
  push:
    branches:
      - 'master'
  pull_request:


jobs:
  test:
    name: 'Node ${{ matrix.node }}'
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node: [12, 14]
    env:
      SKIP_PREFLIGHT_CHECK: true
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-node@v1
      with: 
        node-version: ${{ matrix.node }}
    - name: Install dependencies
      uses: bahmutov/npm-install@v1
      with: 
        install-command: yarn
    - name: Run tests
      run: yarn test:ci