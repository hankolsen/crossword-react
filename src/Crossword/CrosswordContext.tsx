/* eslint-disable @typescript-eslint/no-unused-vars */
import React, {
  createContext,
  FunctionComponent,
  useContext,
  useEffect,
  useState,
} from 'react';
import { useParams } from 'react-router-dom';
import createCrossword, { CrosswordType } from './createCrossword';

type CrosswordContextType = {
  clickHandler: (row: number, column: number) => void;
  crossword?: CrosswordType;
  currentCell: { row: number; column: number };
  isLoading: boolean;
  inputClickHandler: () => void;
  inputHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  keyUpHandler: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  showInput: boolean;
};

const CrosswordContext = createContext<CrosswordContextType | undefined>(
  undefined,
);

const useCrossword = () => {
  const context = useContext(CrosswordContext);
  if (!context) {
    throw new Error('useCrossword must be used within a CrosswordContext');
  }
  return context;
};

const CrosswordProvider: FunctionComponent = ({ children }) => {
  const [crossword, setCrossword] = useState<CrosswordType>();
  const [isLoading, setIsLoading] = useState(true);
  const [showInput, setShowInput] = useState(false);
  const [currentCell, setCurrentCell] = useState({ row: 0, column: 0 });
  const { id } = useParams<{ id: string }>();

  const clickHandler = (row: number, column: number) => {
    if (crossword) {
      setCurrentCell(crossword.cells[row][column]);
    }
    setShowInput(true);
  };

  const inputHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const keyUpHandler = (event: React.KeyboardEvent<HTMLInputElement>) => {};
  const inputClickHandler = () => {};

  useEffect(() => {
    setIsLoading(true);
    createCrossword(id)
      .then((crossword) => setCrossword(crossword))
      .catch()
      .finally(() => setIsLoading(false));
  }, [id]);

  const value = {
    clickHandler,
    crossword,
    currentCell,
    isLoading,
    inputClickHandler,
    inputHandler,
    keyUpHandler,
    showInput,
  };

  return (
    <CrosswordContext.Provider value={value}>
      {children}
    </CrosswordContext.Provider>
  );
};

export { CrosswordProvider, useCrossword };
