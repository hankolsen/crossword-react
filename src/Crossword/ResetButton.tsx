import React from 'react';
import { useCrossword } from './CrosswordContext';

const ResetButton = () => {
  const { isLoading, reset } = useCrossword();

  if (isLoading) {
    return null;
  }

  return (
    <button onClick={reset} type="submit">
      Reset
    </button>
  );
};

export default ResetButton;
