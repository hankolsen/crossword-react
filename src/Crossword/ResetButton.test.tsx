import React from 'react';
import ResetButton from './ResetButton';
import { render, screen } from '../test/test-utils';
import { useCrossword } from './CrosswordContext';

jest.mock('./CrosswordContext');
const mockUseCrossword = useCrossword as jest.Mock;

describe('ResetButton test', () => {
  it('should hide button while loading', () => {
    mockUseCrossword.mockImplementation(() => ({ isLoading: true }));
    const { container } = render(<ResetButton />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });

  it('should render a button', () => {
    mockUseCrossword.mockImplementation(() => ({ isLoading: false }));
    const { container } = render(<ResetButton />);
    expect(container).toMatchSnapshot();
    expect(screen.queryByRole('button')).toBeInTheDocument();
    expect(screen.queryByText(/reset/i)).toBeInTheDocument();
  });
});
