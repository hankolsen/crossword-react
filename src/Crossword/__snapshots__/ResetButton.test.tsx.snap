// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetButton test should hide button while loading 1`] = `
<div>
  <div />
</div>
`;

exports[`ResetButton test should render a button 1`] = `
<div>
  <div>
    <button
      type="submit"
    >
      Reset
    </button>
  </div>
</div>
`;
